Added Dagger/Hilt for Dependency Injection in Tag related screens and surrounding code

danilo04 opened the pull request:
Fixes #1374 

### Fix

Adds support for dependency injection with Dagger/Hilt. Just tag-related screens were migrated. 

### Test

Test tag-related screens

### Review

Only one developer is required to review these changes, anyone can perform the review. Since @hichamboushaba has experience with Hilt, his feedback can be important to improve the integration.

### Release
These changes do not require release notes.

danilo04 committed:
Added dagger hilt dependency and plugins to gradle build files

danilo04 committed:
Added Hilt annotations to Simplenote main application class

Added multiplex dependency since Hilt requires the main application to
inherits from MultiDexApplication.

danilo04 committed:
Added a DI data module to provide data repositories and buckets

The DI data module provides Simperium buckets such as notes and tags.
This module also provides repositories.

danilo04 committed:
Changed AddTagActivity and AddTagViewModel to use DI

danilo04 committed:
Upgraded Hilt dependencies

danilo04 committed:
Added workaround to launchFramgmentInContainer with Hilt

The functions launchFragment and launchFragmentInContainer from
androidx.fragment:fragment-testing are not supported by Hilt. We apply
the workaround described in
https://developer.android.com/training/dependency-injection/hilt-testing

danilo04 committed:
Moved TagsActivity and TagDialogFragment to Hilt

Moved the injection of ViewModels in TagsActivity and TagDialogFragment
to Hilt.

danilo04 committed:
Removed dependency for fragment-testing since it is not needed

danilo04 committed:
Added build cache configuration

danilo04 committed:
Rollback change to add cache in gradle

loremattei committed:
Fix out of memory error on CI

danilo04 committed:
Merge pull request #1433 from Automattic/fix/out-of-memory-error

Fix out of memory error on CI

hichamboushaba commented on Simplenote/build.gradle:
```diff
@@ -36,6 +38,7 @@ android {
         versionCode 138
         minSdkVersion 23
         targetSdkVersion 29
+        multiDexEnabled true
```
Can you explain why we need this? I think that when minSdkVersion is higher than 21, multi-dex is enabled by default, and we don't need anything from our side (source https://developer.android.com/studio/build/multidex.html#mdex-on-l)

danilo04 replied:
Good catch, thanks! It was mostly based on the `Hilt` documentation that includes it. I have removed this line in `build.gradle` and changed `Simplenote` to inherit from `Application`. Fixed in a72e8cec053b5372338317186403671580695f80.

hichamboushaba commented on Simplenote/src/androidTest/java/com/automattic/simplenote/utils/HiltUtils.kt:
```diff
@@ -0,0 +1,43 @@
+ * As a workaround, use this function that is equivalent. It requires you to add
+ * [HiltTestActivity] in the debug folder and include it in the debug AndroidManifest.xml file
+ * as can be found in this project. This function focus on launching DialogFragment
+ */
```
np, personally I would update this comment to mention also the source of the solution https://developer.android.com/training/dependency-injection/hilt-testing#launchfragment, it would make understanding it easier, WDYT?

danilo04 replied:
Sure!. Fixed in 9b1e046169e12a56c8aaa725e733004e65d11583.

hichamboushaba commented on Simplenote/src/main/java/com/automattic/simplenote/di/DataModule.kt:
```diff
@@ -0,0 +1,29 @@
+    @Provides
+    fun providesNotesBucket(@ApplicationContext appContext: Context): Bucket<Note> = (appContext as Simplenote).notesBucket
+
+    @Provides
+    fun providesTagsRepository(tagsBucket: Bucket<Tag>, notesBucket: Bucket<Note>): TagsRepository =
+        SimperiumTagsRepository(tagsBucket, notesBucket)
```
np, I would change this to use `@Binds` instead of `@Provides`, this is because binding will decouple the injection from the constructor's implementation, and if the number of arguments needed for the constructor changes, we won't need to update in the module itself.

A patch to make my suggestion clearer 🙂 
```
Index: Simplenote/src/main/java/com/automattic/simplenote/di/DataModule.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- Simplenote/src/main/java/com/automattic/simplenote/di/DataModule.kt	(revision 31029a7742cf2529eed6e65e50fb7048664c748c)
+++ Simplenote/src/main/java/com/automattic/simplenote/di/DataModule.kt	(date 1628528283397)
@@ -7,6 +7,7 @@
 import com.automattic.simplenote.repositories.SimperiumTagsRepository
 import com.automattic.simplenote.repositories.TagsRepository
 import com.simperium.client.Bucket
+import dagger.Binds
 import dagger.Module
 import dagger.Provides
 import dagger.hilt.InstallIn
@@ -15,15 +16,15 @@
 
 @Module
 @InstallIn(SingletonComponent::class)
-object DataModule {
-
-    @Provides
-    fun providesTagsBucket(@ApplicationContext appContext: Context): Bucket<Tag> = (appContext as Simplenote).tagsBucket
+abstract class DataModule {
+    companion object {
+        @Provides
+        fun providesTagsBucket(@ApplicationContext appContext: Context): Bucket<Tag> = (appContext as Simplenote).tagsBucket
 
-    @Provides
-    fun providesNotesBucket(@ApplicationContext appContext: Context): Bucket<Note> = (appContext as Simplenote).notesBucket
+        @Provides
+        fun providesNotesBucket(@ApplicationContext appContext: Context): Bucket<Note> = (appContext as Simplenote).notesBucket
+    }
 
-    @Provides
-    fun providesTagsRepository(tagsBucket: Bucket<Tag>, notesBucket: Bucket<Note>): TagsRepository =
-        SimperiumTagsRepository(tagsBucket, notesBucket)
-}
\ No newline at end of file
+    @Binds
+    abstract fun bindsTagsRepository(repository: SimperiumTagsRepository): TagsRepository
+}
Index: Simplenote/src/main/java/com/automattic/simplenote/repositories/SimperiumTagsRepository.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- Simplenote/src/main/java/com/automattic/simplenote/repositories/SimperiumTagsRepository.kt	(revision 31029a7742cf2529eed6e65e50fb7048664c748c)
+++ Simplenote/src/main/java/com/automattic/simplenote/repositories/SimperiumTagsRepository.kt	(date 1628528399499)
@@ -17,8 +17,9 @@
 import kotlinx.coroutines.flow.callbackFlow
 import kotlinx.coroutines.flow.flowOn
 import kotlinx.coroutines.withContext
+import javax.inject.Inject
 
-class SimperiumTagsRepository(
+class SimperiumTagsRepository @Inject constructor(
         private val tagsBucket: Bucket<Tag>,
         private val notesBucket: Bucket<Note>
 ) : TagsRepository {
```

danilo04 replied:
That is a great patch @hichamboushaba! I have applied it in ac26d345b2a5438e69b8b8dd6cda827b467a4070.

hichamboushaba reviewed (commented):
Looks nice @danilo04 👍, I left some questions/suggestions, but nothing major 🙂

danilo04 committed:
Remove dependency to multi dex since it is included by default

Since minSdkVersion 21, multidex is enabled by default, so we do not
need the multidex library.

danilo04 committed:
Merge remote-tracking branch 'origin/issue/1374-add-dependency-injection-lib' into issue/1374-add-dependency-injection-lib

danilo04 committed:
Add source page in HiltUtils to mention where it comes from

danilo04 committed:
Use @Binds instead of @Provides for SimperiumRepository instance

Use @Binds instead of @Provides because binding will decouple the
injection from the constructor's implementation, and if the number of
arguments needed for the constructor changes, we won't need to update
in the module itself.

danilo04 commented:
Thanks for the suggestions @hichamboushaba!

hichamboushaba reviewed (approved):
Thanks for the changes @danilo04, nice work 👍
